cranelift/codegen/src/isa/zkasm/abi.rs

```diff
@@ -511,21 +518,12 @@ impl ABIMachineSpec for Riscv64MachineDeps {
         frame_layout: &FrameLayout,
     ) -> SmallVec<[Self::I; 16]> {
         let mut insts = SmallVec::new();
-        // Adjust the stack pointer downward for clobbers and the function fixed
+        // Adjust the stack pointer upward for clobbers and the function fixed
         // frame (spillslots and storage slots).
         let stack_size = frame_layout.fixed_frame_storage_size + frame_layout.clobber_size;
         // Each stack slot is 64 bit and can fit a u8 value.
+        // TODO(nagisa): What this comment above _means_??
```

Comment:
It's accounting for the fact that the stack is addressed in words, 8 bytes each. Concretely, I was trying to address the fact that when we put a register on the stack, we would advance the stack by 8 slots instead of 1.
The division by 8 that I did is not the correct way though, as it doesn't take into account that all reads must be word-aligned and so pushing two u32s would likely yield an incorrect shift by only 1 slot.
I'm not quite sure what is the right fix, though...

Reply:
Is there a reason to operate in 64-bit stack slots, when the underlying machine works with byte-addressed stack? I saw mentions in the documentation that the machine tear the unaligned loads and stores, but dealing with that is a concern of setting up the alignment of types right.

Reply:
So, uh, memory operation wise we have, as far as I understand the following operations:

* `MLOAD`, `MSTORE` – byte addressable, reads/writes a full 256-bit register, unspecified how it handles unaligned addresses;
* `MEM_ALIGN_RD`, `MEM_ALIGN_WR` – byte addressable, reads/writes a full 256-bit register, address can be byte-granular;
* `MEM_ALIGN_WR8` – byte addresssable, reads/writes a byte within the specified location of the specified register.

`MEM_ALIGN_WR8` seems particularly poor fit for writing out values that aren't a byte (provided that we'd have to issue 2 of these for 16-bit writes, 4 of these for 32-bit writes and so on.) So from what I can tell our only option is to align all values to 256-bits right now and stomach the waste of space that this will result in when handling small values.